solidity/contracts/converter/BancorConverter.sol

```diff
@@ -402,9 +419,7 @@ contract BancorConverter is IBancorConverter, SmartTokenController, ContractRegi
         validReserve(_reserveToken)
         returns (uint256)
     {
-        if (_reserveToken == IERC20Token(0))
-            return address(this).balance;
-        return _reserveToken.balanceOf(this);
+        return reserveBalances[_reserveToken];

```

Comment:
See my answer about re. getReserveBalance.

Other than that, I imagine we can remove the check from some of these functions but how are we gonna handle getPurchaseReturn? That function has to be able to indicate if the input is invalid, it can't just return 0 since that's a valid return amount (theoretically speaking).

Reply:
Whoever calls `getPurchaseReturn`, should verify beforehand that the reserve is valid. I imagine that's already the case anyway (either that, or the reserve is known to be valid, as is the case when this function is called from one of our conversion functions (`convertInternal` or whatever)).

Reply:
`getPurchaseReturn` is called from two places today -
- `getReturn`
- `buy`

Neither of these verify the reserve address - they actually rely on `getPurchaseReturn` for that - do you suggest we remove it from the internal function and put it externally in both functions?

Reply:
@yudilevi:
Function `getReturn` is read-only. I don't see a problem with it returning 0, because that's the expected return when a reserve is not valid.
Function `buy` can (and maybe should) validate the reserve beforehand, and IMO it should rely on `getPurchaseReturn` to do that. BTW, it doesn't really matter if this function doesn't validate the reserve, because it does `require(amount != 0)` immediately after calling `getPurchaseReturn`.
But the issue is general not only to `getPurchaseReturn`, but also with several other functions as I have mentioned above.
IMO, we should validate the reserve only in write functions (and even there, only if it seems necessary).